### ISSUE TITLE
Me again UwU - with some good and some bad news (development on port 3000 works, deploy fails)

### ISSUE THREAD

**Issue by GlassedSilver** [danbooru/danbooru#4264]
Hey guys! So I got some amazing news. The LOVELY set up instructions in the wiki have greatly helped me get up to speed and so far I've already filled my local booru in a dev setup with a few images to make sure everything moves along nicely over a couple of updates etc... Good old software testing.

Unfortunately it seems to get custom settings using automated overrides (like Pixiv and other passwords) with the proper file for it I need to run the deployment I take it? Either way, when I try to deploy this happens:

```
danbooru@ahri-vm-danbooru:~/danbooru$ bundle exec cap production deploy
00:00 rbenv:validate
      rbenv: rbenv_ruby is not set; ruby version will be defined by the remote hosts via rbenv
00:00 deploy:prepare_tree
      01 git fetch origin
    ✔ 01 danbooru@localhost 0.994s
      02 git checkout master
      02 Already on 'master'
      02 Your branch is up to date with 'origin/master'.
    ✔ 02 danbooru@localhost 0.012s
      Pulling from master
      03 git pull origin master
      03 From https://github.com/GlassedSilver/danbooru
      03  * branch                master     -> FETCH_HEAD
      03 Already up to date.
    ✔ 03 danbooru@localhost 0.819s
00:01 git:wrapper
      01 mkdir -p /tmp
#<Thread:0x00005615547706e0@/home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/sshkit-1.20.0/lib/sshkit/runners/parallel.rb:10 run> terminated with exception (report_on_exception is true):
Traceback (most recent call last):
        26: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/sshkit-1.20.0/lib/sshkit/runners/parallel.rb:12:in `block (2 levels) in execute'
        25: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/sshkit-1.20.0/lib/sshkit/backends/abstract.rb:31:in `run'
        24: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/sshkit-1.20.0/lib/sshkit/backends/abstract.rb:31:in `instance_exec'
        23: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/capistrano-3.11.2/lib/capistrano/scm/tasks/git.rake:8:in `block (3 levels) in eval_rakefile'
        22: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/sshkit-1.20.0/lib/sshkit/backends/abstract.rb:80:in `execute'
        21: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/sshkit-1.20.0/lib/sshkit/backends/abstract.rb:148:in `create_command_and_execute'
        20: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/sshkit-1.20.0/lib/sshkit/backends/abstract.rb:148:in `tap'
        19: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/sshkit-1.20.0/lib/sshkit/backends/abstract.rb:148:in `block in create_command_and_execute'
        18: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/sshkit-1.20.0/lib/sshkit/backends/netssh.rb:130:in `execute_command'
        17: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/sshkit-1.20.0/lib/sshkit/backends/netssh.rb:177:in `with_ssh'
        16: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/sshkit-1.20.0/lib/sshkit/backends/connection_pool.rb:63:in `with'
        15: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/sshkit-1.20.0/lib/sshkit/backends/connection_pool.rb:63:in `call'
        14: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/net-ssh-5.2.0/lib/net/ssh.rb:250:in `start'
        13: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/net-ssh-5.2.0/lib/net/ssh/authentication/session.rb:71:in `authenticate'
        12: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/net-ssh-5.2.0/lib/net/ssh/authentication/session.rb:71:in `each'
        11: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/net-ssh-5.2.0/lib/net/ssh/authentication/session.rb:85:in `block in authenticate'
        10: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/net-ssh-5.2.0/lib/net/ssh/authentication/methods/publickey.rb:19:in `authenticate'
         9: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/net-ssh-5.2.0/lib/net/ssh/authentication/key_manager.rb:119:in `each_identity'
         8: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/net-ssh-5.2.0/lib/net/ssh/authentication/key_manager.rb:119:in `each'
         7: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/net-ssh-5.2.0/lib/net/ssh/authentication/key_manager.rb:122:in `block in each_identity'
         6: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/net-ssh-5.2.0/lib/net/ssh/authentication/methods/publickey.rb:20:in `block in authenticate'
         5: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/net-ssh-5.2.0/lib/net/ssh/authentication/methods/publickey.rb:62:in `authenticate_with'
         4: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/net-ssh-5.2.0/lib/net/ssh/authentication/key_manager.rb:142:in `sign'
         3: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/net-ssh-5.2.0/lib/net/ssh/key_factory.rb:44:in `load_private_key'
         2: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/net-ssh-5.2.0/lib/net/ssh/key_factory.rb:53:in `load_data_private_key'
         1: from /home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/net-ssh-5.2.0/lib/net/ssh/key_factory.rb:202:in `classify_key'
/home/danbooru/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/net-ssh-5.2.0/lib/net/ssh/authentication/ed25519_loader.rb:21:in `raiseUnlessLoaded': OpenSSH keys only supported if ED25519 is available (NotImplementedError)
net-ssh requires the following gems for ed25519 support:
 * ed25519 (>= 1.2, < 2.0)
 * bcrypt_pbkdf (>= 1.0, < 2.0)
See https://github.com/net-ssh/net-ssh/issues/565 for more information
Gem::LoadError : "ed25519 is not part of the bundle. Add it to your Gemfile."
(Backtrace restricted to imported tasks)
cap aborted!
NotImplementedError: OpenSSH keys only supported if ED25519 is available
net-ssh requires the following gems for ed25519 support:
 * ed25519 (>= 1.2, < 2.0)
 * bcrypt_pbkdf (>= 1.0, < 2.0)
See https://github.com/net-ssh/net-ssh/issues/565 for more information
Gem::LoadError : "ed25519 is not part of the bundle. Add it to your Gemfile."

Tasks: TOP => deploy:check => git:check => git:wrapper
(See full trace by running task with --trace)
The deploy has failed with an error: OpenSSH keys only supported if ED25519 is available
net-ssh requires the following gems for ed25519 support:
 * ed25519 (>= 1.2, < 2.0)
 * bcrypt_pbkdf (>= 1.0, < 2.0)
See https://github.com/net-ssh/net-ssh/issues/565 for more information
Gem::LoadError : "ed25519 is not part of the bundle. Add it to your Gemfile."


** DEPLOY FAILED
** Refer to log/capistrano.log for details. Here are the last 20 lines:


  INFO [944198a6] Running /usr/bin/env git fetch origin as danbooru@localhost

 DEBUG [944198a6] Command: ( export RBENV_ROOT="$HOME/.rbenv" RBENV_VERSION="" ; /usr/bin/env git fetch origin )

  INFO [944198a6] Finished in 0.994 seconds with exit status 0 (successful).

 DEBUG [59f761e6] Running /usr/bin/env git diff master --shortstat as danbooru@localhost

 DEBUG [59f761e6] Command: ( export RBENV_ROOT="$HOME/.rbenv" RBENV_VERSION="" ; /usr/bin/env git diff master --shortstat )

 DEBUG [59f761e6] Finished in 0.008 seconds with exit status 0 (successful).

  INFO [9a0b4d83] Running /usr/bin/env git checkout master as danbooru@localhost

 DEBUG [9a0b4d83] Command: ( export RBENV_ROOT="$HOME/.rbenv" RBENV_VERSION="" ; /usr/bin/env git checkout master )

 DEBUG [9a0b4d83]       Already on 'master'

 DEBUG [9a0b4d83]       Your branch is up to date with 'origin/master'.

  INFO [9a0b4d83] Finished in 0.012 seconds with exit status 0 (successful).

  INFO Pulling from master

  INFO [392093a5] Running /usr/bin/env git pull origin master as danbooru@localhost

 DEBUG [392093a5] Command: ( export RBENV_ROOT="$HOME/.rbenv" RBENV_VERSION="" ; /usr/bin/env git pull origin master )

 DEBUG [392093a5]       From https://github.com/GlassedSilver/danbooru

 DEBUG [392093a5]        * branch                master     -> FETCH_HEAD

 DEBUG [392093a5]       Already up to date.

  INFO [392093a5] Finished in 0.819 seconds with exit status 0 (successful).

  INFO [ca5247c9] Running /usr/bin/env mkdir -p /tmp on localhost

 DEBUG [ca5247c9] Command: ( export RBENV_ROOT="$HOME/.rbenv" RBENV_VERSION="" ; /usr/bin/env mkdir -p /tmp )

danbooru@ahri-vm-danbooru:~/danbooru$
```
If more info is needed like the log itself (where would it be for capistrano? I know we looked for other logs last time around, but cap wasn't one of them, well new issue too I guess...)

Anyhow, already I've gotten so much closer to my goal and I cannot be thankful enough!

Bonus question: (please tell me if I should open a new issue for this, I'll merrily do it)

How can I disable usage of Cloudflare? It's an entirely local setup, so if I could have everything work on its own, all the better. I'm also interested in mocking Amazon SQS if that gives me a working IQDB service. Is Amazon merely used to optimize/load balance queries or is it actually executing code and hard to mock in a way that renders services dependent on it useful?

Sorry if I'm being a bit of an annoyance, but now I've really tasted blood and am eager to get to the finish line. :D What an amazing feeling to see your own booru serving you, heck even through mobile clients and to know that you're hosting this on your own terms, meta data included and easy browsing... I digress. Just... thank you!

**Comment by thenamewastaken** [comment]
You have to add the missing gems to the Gemfile:
```
gem 'bcrypt_pbkdf'
gem 'ed25519', '< 2.0'
```

**Comment by evazion** [comment]
You're using a ED25519 format SSH key. You can either switch to an RSA key (`ssh-keygen -t rsa`) or add the above gems to your Gemfile. I don't want to add these dependencies myself since they're not needed in most cases.

If you're just setting up a personal instance, you don't need to use Capistrano. You can just install everything then run `bin/rails server` directly. Capistrano is meant for automating the process of deploying the site to a remote server.

> How can I disable usage of Cloudflare? It's an entirely local setup, so if I could have everything work on its own, all the better. 

Cloudflare is optional, you can skip it. It's only need if you're using Cloudflare and you want to purge Cloudflare's cache when an image is replaced or deleted.

> I'm also interested in mocking Amazon SQS if that gives me a working IQDB service. Is Amazon merely used to optimize/load balance queries or is it actually executing code and hard to mock in a way that renders services dependent on it useful?

When a post is uploaded, a message is sent to SQS. A separate process in the IQDBS service receives these messages and adds the new images to IQDB. All we do is basic sending and receiving of messages, so in theory mocking SQS should be relatively straightforward.

**Comment by GlassedSilver** [comment]
Okay I scrapped trying to run this in prod, because I have grown confident with the dev branch being not only stable, but also easily maintained from my limited experience. I was even able to fix a borked set of dependencies after the VM got interrupted ungracefully.

So far, so good! :) Really happy with how it's all running, still just need to figure out how to **mock SQS**, because I have absolutely zero experience with it.

Having IQDB would be really helpful, because that I think is what would be used to inform me if I'm trying to add an image that's already on the server or one that looks similar, right? Or are there further dependencies I need to be aware of? (just SQS mocked and IQDB to my understanding)

**My question about cloudflare** was about the fact that if I don't allow these entries in uMatrix:
```
192.168.2.229 cdnjs.cloudflare.com script allow
192.168.2.229 cloudflare.com * allow
```
The pages don't load properly. That worries me a little for two key reasons: a) privacy b) I want this instance not to depend on anything external other than for obvious reasons like pulling from Pixiv needs an internet connection obviously, but viewing, browsing, etc... should all be independent of any third-party.

Thirdly, what do I need to do to make **pulling from Twitter** work properly? It does pull the artist, but it will not pull the image (format not supported, I think it doesn't even detect the image file of the post) and it also doesn't pull the comment of the post. I am using the bookmarklet as well as trying manually pasting the Twitter post's URL into the source field.

Fourthly: I cannot emphasize enough how much you guys have helped me and made me find peace and comfort in knowing that I can finally store all those pictures on my terms, properly tagged, sourced, etc... It's all I ever could ask for and it's wonderful. I have already started submitting images and tags and other edits to official danbooru, because that's the work I can do right now for everyone else. I will surely do a lot more as I eventually move to archiving pics that aren't on danbooru yet. Furthermore I want to know if there is any way to directly sponsor development of danbooru beyond upgrading my account on danbooru? I'm having problems with Stripe atm, but maybe it's just my card not playing ball with international transactions, although that SHOULD work. I'll look into it another day, just wanna say that I look forward to giving back to development and the community as a whole. <3 :)
